File: sources/global/util/src/main/java/com/tars/util/net/http/HttpHub.java
package com.tars.util.net.http;

import org.apache.http.client.methods.HttpUriRequest;
import org.apache.http.client.methods.RequestBuilder;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.apache.http.util.EntityUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;

import util.Option;

import static util.global.ExceptionHandlingKt.executeAndGetSilent;
import static util.global.ExceptionHandlingKt.wtf;



/**
 * HttpHub provides interface for interacting with http protocol.
 */
public final class HttpHub {

  private static final Logger log = LoggerFactory.getLogger("http");
  private CloseableHttpClient hc;

  // lifecycle

  public HttpHub() {
    log.info("init");
    hc = HttpClients.createDefault();
  }

  public void release() {
    log.info("shutdown");
    try {
      hc.close();
    } catch (IOException e) {
      wtf(e);
    }
  }

  // interface

  public Option<String> getString(RequestBuilder requestBuilder) {
    HttpUriRequest request = requestBuilder.build();
    log.debug("requesting {}", request);

    return executeAndGetSilent(
        () -> EntityUtils.toString(hc.execute(request).getEntity())
    );
  }

  public Option<String> getString(HttpUriRequest request) {
    log.debug("requesting {}", request);

    return executeAndGetSilent(
        () -> EntityUtils.toString(hc.execute(request).getEntity())
    );
  }

}

File: sources/global/util/src/main/kotlin/util/network/http/HttpHub.kt
package util.network.http

import org.apache.http.client.methods.HttpUriRequest
import org.apache.http.client.methods.RequestBuilder
import org.apache.http.impl.client.HttpClients
import org.apache.http.util.EntityUtils
import util.Option
import util.maid
import util.global.executeAndGetSilent
import util.global.logger


/**
 * HttpHub provides interface for interacting with http protocol.
 */
class HttpHub {
    internal val log = logger("http")
    internal val hc = HttpClients.createDefault().apply { maid.internalAdd({ this.close() }, 1, "http-client") }

    // interface

    fun get(url: String): Option<String> {
        return executeAndGetSilent { EntityUtils.toString(hc.execute(RequestBuilder.get(url).build()).entity) }
    }

    fun getString(requestBuilder: RequestBuilder): Option<String> {
        return getString(requestBuilder.build())
    }

    fun getString(request: HttpUriRequest): Option<String> {
        log.debug { "requesting ${request.uri}" }
        return executeAndGetSilent { EntityUtils.toString(hc.execute(request).entity) }
    }

}
